### PR TITLE
chore(core): improve `LogNamespace` helper functions

### PR DESCRIPTION
**Pull request by fuchsnj**
The most important change here is the addition of the `LegacyKey` enum, which allows the `LogNamespace` helper functions to differentiate between keys that should `try_insert` vs `insert`. Previously they would only `try_insert`, but this didn't match the behavior of some components.

Along with these changes, paths related to type definitions were switched to `OwnedValueType`. This has a few benefits
- This migrates away from the old path code
- It removes an "impl trait" argument, which can be annoying to use when the argument is optional (You have to add a turbofish when the `None` variant is used)

### CODE REVIEW

File: src/sources/amqp.rs [hunk 1]
```diff
@@ -213,22 +214,28 @@ fn populate_event(
...
+        AmqpSourceConfig::NAME,
         log,
-        keys.routing_key_field,
+        Some(LegacyKey::InsertIfEmpty(keys.routing_key_field)),
         "routing",
         keys.routing.to_string(),
     );
 
     log_namespace.insert_source_metadata(
         "amqp",
```
Comment (spencergilbert) [inline src/sources/amqp.rs:225]:
```suggestion
        AmqpSourceConfig::NAME,
```

File: src/sources/amqp.rs [hunk 2]
```diff
@@ -213,22 +214,28 @@ fn populate_event(
...
         log,
-        keys.exchange_key,
+        Some(LegacyKey::InsertIfEmpty(keys.exchange_key)),
         "exchange",
         keys.exchange.to_string(),
     );
 
-    log_namespace.insert_source_metadata("amqp", log, keys.offset_key, "offset", keys.delivery_tag);
+    log_namespace.insert_source_metadata(
+        "amqp",
```
Comment (spencergilbert) [inline src/sources/amqp.rs:233]:
```suggestion
        AmqpSourceConfig::NAME,
```

File: src/sources/datadog_agent/logs.rs [hunk 3]
```diff
@@ -111,55 +112,55 @@ pub(crate) fn decode_log_body(
...
                                 log,
-                                path!("ddtags"),
+                                Some(LegacyKey::InsertIfEmpty(path!("ddtags"))),
                                 path!("ddtags"),
                                 ddtags.clone(),
                             );
 
                             namespace.insert_vector_metadata(
                                 log,
-                                path!(source.log_schema_source_type_key),
```
Comment (spencergilbert) [inline src/sources/datadog_agent/logs.rs:156]: Is it just not necessary to macro this, the type already matches?

File: src/sources/datadog_agent/mod.rs [hunk 4]
```diff
@@ -194,51 +195,44 @@ impl SourceConfig for DatadogAgentConfig {
             .decoding
             .schema_definition(global_log_namespace.merge(self.log_namespace))
             .with_source_metadata(
-                self.get_component_name(),
-                Some("message"),
-                "message",
-                Kind::bytes(),
-                Some("message"),
```
Comment (spencergilbert) [inline src/sources/datadog_agent/mod.rs:201]: Is the dd decoder properly handling this type definition?

**Review by spencergilbert — COMMENTED** [review]
Only skimmed the datadog test changes...

File: src/sources/datadog_agent/logs.rs (hunk 3 above)
Comment (fuchsnj) [reply src/sources/datadog_agent/logs.rs:156]: This was previously a mistake. The log schema keys are supposed to be full keys (may contain multiple segments). The macro was forcing it to be a single segment. (It's still being passed in as a string now, but that will improve in the future)

File: src/sources/datadog_agent/mod.rs (hunk 4 above)
Comment (fuchsnj) [reply src/sources/datadog_agent/mod.rs:201]: Yes, this is handled by the decoder.

File: src/sources/amqp.rs [hunk 5]
```diff
@@ -213,22 +214,28 @@ fn populate_event(
     );
 
     log_namespace.insert_source_metadata(
-        "amqp",
+        AmqpSourceConfig::NAME,
         log,
-        keys.routing_key_field,
+        Some(LegacyKey::InsertIfEmpty(keys.routing_key_field)),
         "routing",
```
Comment (spencergilbert) [inline src/sources/amqp.rs:220]: Should these theoretically be `path!`?

**Review by spencergilbert — APPROVED** [review]

File: src/sources/amqp.rs (hunk 5 above)
Comment (fuchsnj) [reply src/sources/amqp.rs:220]: No. These should be parsed as potentially nested paths. The config for `routing_key_field` will be switched to `OwnedTargetPath` in the near future, so it won't be using "String" paths then.

Comment (spencergilbert) [reply src/sources/amqp.rs:220]: Wouldn't that make [this](https://github.com/vectordotdev/vector/blob/master/src/sources/aws_s3/sqs.rs#L527) wrong?

Comment (fuchsnj) [reply src/sources/amqp.rs:220]: `path!("bucket")` is a single segment path, so the macro handles that correctly. If the string literal was a path with multiple segments (eg: `bucket.foo`) than just passing the whole thing to the `path` macro would be incorrect (`path!("bucket.fo")`, instead you would need to manually split it and call it like `path!("bucket", "foo")`

Comment (spencergilbert) [reply src/sources/amqp.rs:220]: Ah, yeah because this is user submitted - we can't be sure. Thanks!

**Pull request merged**